### PR TITLE
[CAS-446] MessageInputView improvements

### PR DESCRIPTION
**Pull request by samiuelson**
### Description

Next iteration over MessageInputView:
1. Allow to send plain message
2. Allow to send message to thread
3. "send also to channel" option

### Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [ ] Changelog updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [x] Reviewers added

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt [hunk 1]
```diff
@@ -430,3 +434,7 @@ public interface MessageSendHandler {
 
     public fun editMessage(oldMessage: Message, newMessageText: String)
 }
+
+public fun interface OnMessageSendButtonClickListener {
```
Comment (zsmb13) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt:438]: This seems to belong strictly to the class above as well, let's nest these, it adds logical scoping

**Review by zsmb13 — COMMENTED** [review]
Looks good functionally (though the message list jumps around after sending messages, a separate issue).

Left a bunch of small notes and questions.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/BaseMessageItemViewHolder.kt [hunk 2]
```diff
@@ -4,6 +4,7 @@ import android.view.View
 import androidx.recyclerview.widget.RecyclerView
 import com.getstream.sdk.chat.adapter.MessageListItem
 
-public abstract class BaseMessageItemViewHolder<T : MessageListItem>(itemView: View) : RecyclerView.ViewHolder(itemView) {
+public abstract class BaseMessageItemViewHolder<T : MessageListItem>(itemView: View) :
+    RecyclerView.ViewHolder(itemView) {
```
[the remark this answers is not in the document]
Comment (samiuelson) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/BaseMessageItemViewHolder.kt:8]: Something weird happened. I thought it was rebase related, but maybe it's on my end. I'll look into this.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt [hunk 3]
```diff
@@ -322,4 +366,67 @@ public class MessageInputView : ConstraintLayout {
...
+    public sealed class ChatMode {
+        public object DirectChat : ChatMode()
+        public object GroupChat : ChatMode()
+    }
+}
+
+public interface TypeListener {
+    public fun onKeystroke()
+    public fun onStopTyping()
+}
```
[the remark this answers is not in the document]
Comment (samiuelson) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt]: Removed

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt [hunk 4]
```diff
@@ -322,4 +366,67 @@ public class MessageInputView : ConstraintLayout {
...
+    public sealed class InputMode {
+        public object Normal : InputMode()
+        public data class Thread(val parentMessage: Message) : InputMode()
+        public data class Edit(val oldMessage: Message) : InputMode()
+    }
+
+    public sealed class ChatMode {
+        public object DirectChat : ChatMode()
+        public object GroupChat : ChatMode()
+    }
```
[the remark this answers is not in the document]
Comment (samiuelson) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt]: turned to enum

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/BaseMessageItemViewHolder.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (samiuelson) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/BaseMessageItemViewHolder.kt:8]: Yep, my bad. Seems I have an IDE-level pre-commit auto format task that resulted in committing changes I wasn't aware of. Bad stuff.

[the remark this answers is not in the document]
Comment (samiuelson) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/messages/adapter/BaseMessageItemViewHolder.kt:8]: I've reverted the formatting-related changes here. Left them in other files where they seem to make more sense.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt [hunk 5]
```diff
@@ -78,6 +99,36 @@ public class MessageInputView : ConstraintLayout {
...
+                    }
+                }
+            }
+
+            messageText = ""
+        }
+    }
+
+    private fun configSendAlsoToChannelCheckbox() {
+        binding.sendAlsoToChannel.isVisible = inputMode is InputMode.Thread
```
[the remark this answers is not in the document]
Comment (samiuelson) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/textinput/MessageInputView.kt]: Added.

**Comment by samiuelson** [comment]
> Looks good functionally (though the message list jumps around after sending messages, a separate issue).
> 
> Left a bunch of small notes and questions.

Yep, we're using old `MessageListView` implementation for the time being.

Thanks for rev!

File: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt [hunk 6]
```diff
@@ -45,11 +46,57 @@ class ChatFragment : Fragment() {
...
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         headerViewModel.bindView(binding.header, viewLifecycleOwner)
-        messageListViewModel.bindView(binding.messageList, viewLifecycleOwner)
-        inputListViewModel.bindView(binding.input, viewLifecycleOwner)
+        initMessagesViewModel()
+        initMessageInputViewModel()
+        configureBackButtonHandlinig()
+    }
 
+    private fun configureBackButtonHandlinig() {
```
Comment (zsmb13) [inline stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt:54]:
```suggestion
    private fun configureBackButtonHandling() {
```

**Review by zsmb13 — APPROVED** [review]
LGTM

**Pull request merged**